### PR TITLE
Update math extension

### PR DESCRIPTION
**Pull request by dvdvideo1234**
Added: math.Sign

### CODE REVIEW

File: garrysmod/lua/includes/extensions/math.lua [hunk 1]
```diff
@@ -1,3 +1,11 @@
+--[[---------------------------------------------------------
+	Name: Sign( num )
+	Desc: Takes the sign of a number, 0 when not relevant
+------------------------------------------------------------]]
+function math.Sign(num)
+  local num = tonumber(num) or 0
```
Comment (AMD-NICK) [inline garrysmod/lua/includes/extensions/math.lua]: I think it's unnecessary

**Comment by thegrb93** [comment]
It has its uses. I think if then else is better than all this arithmetic though.
```
if a>0 then
    return 1
elseif a<0 then
    return -1
else
    return 0
end
```

**Comment by robotboy655** [comment]
I think this whole PR is unnecessary.

**Comment by dvdvideo1234** [comment]
Why.. There is no Sign function. I have to make it as a library every time

**Comment by bigdogmat** [comment]
> I think it will crash when u get the sign of a string as you cannot compare string to number.

It'll just error, either way I think the point is barely any of the other math functions use tonumber on arguments. If someone is using a string input and is trying to use it as a number, they should purposely convert it before hand.

**Comment by thegrb93** [comment]
Math functions don't need type safety imo.

**Comment by AMD-NICK** [comment]
> It has its uses. I think if then else is better than all this arithmetic though.

this should be better
```lua
function math.Sign(num)
   eturn (num > 0 and 1) or (num < 0 and -1) or 0
end
```

**Comment by dvdvideo1234** [comment]
What happens when you call it like this:

local a = math.Sin("50")
print("The result is "..a)

**Comment by bigdogmat** [comment]
> If someone is using a string input and is trying to use it as a number, they should purposely convert it before hand.

**Comment by dvdvideo1234** [comment]
@bigdogmat Ok then, removed the tonumber conversion

**Comment by thegrb93** [comment]
local a = math.Sign(tonumber("50"))?

**Comment by thegrb93** [comment]
My version is the fastest one.
```
require 'xbench'
--jit.off()

function math.Sign(num)
	return (num ~= 0) and (num / math.abs(num)) or 0
end

function math.Sign2(num)
	return (num > 0 and 1) or (num < 0 and -1) or 0
end

function math.Sign3(a)
	if a>0 then
		return 1
	elseif a<0 then
		return -1
	else
		return 0
	end
end

xbench.Compare({
    Old = function()
        math.Sign( 0.15 )
        math.Sign( -0.15 )
    end,
    New = function()
        math.Sign2( 0.15 )
        math.Sign2( -0.15 )
    end,
}, 1000000)

xbench.Compare({
    Old = function()
        math.Sign( 0.15 )
        math.Sign( -0.15 )
    end,
    New = function()
        math.Sign3( 0.15 )
        math.Sign3( -0.15 )
    end,
}, 1000000)
```
OLD: 0.11174105958327
NEW: 0.07191243159474
OLD: 0.10693967295873
NEW: 0.067574400211555

**Comment by dvdvideo1234** [comment]
Nice, thanks for testing this for me ;)

**Comment by AMD-NICK** [comment]
@thegrb93 fraud!11! :D
image

anyway my method looks better

```lua
require("xbench")

function math.Sign1(num)
	return (num ~= 0) and (num / math.abs(num)) or 0
end

function math.Sign2(num)
	return (num > 0 and 1) or (num < 0 and -1) or 0
end

function math.Sign3(a)
	if a > 0 then
		return 1
	elseif a < 0 then
		return -1
	else
		return 0
	end
end

xbench.Compare({
	Sign1 = function()
		math.Sign1( 0.15 )
		math.Sign1( 1 )
		math.Sign1( -0.15 )
	end,
	Sign2 = function()
		math.Sign2( 0.15 )
		math.Sign2( 0 )
		math.Sign2( -0.15 )
	end,
	Sign3 = function()
		math.Sign3( 0.15 )
		math.Sign3( 0 )
		math.Sign3( -0.15 )
	end,
}, 100000000)

-- SIGN2: 0.029420669774481
-- SIGN1: 0.034826290601643
-- SIGN3: 0.029995217636326
```

**Comment by dvdvideo1234** [comment]
Bit it will look longer like one-liner

'''
function math.Sign3(a)
    if a > 0 then return 1 elseif a < 0 then return -1 else return 0 end
end
'''

**Comment by Kefta** [comment]
My benchmarks were inconclusive because the difference ended up being purely based on processor tick time and no trial could prove which is the "fastest." I would say just go with Sign2 because it's the cleanest and doesn't call an outside method.

**Comment by dvdvideo1234** [comment]
I bench my programs with my own code, that compares then for answer consistency :D

https://github.com/dvdvideo1234/ZeroBraineProjects/blob/master/dvdlualib/common.lua#L490-L553

**Comment by dvdvideo1234** [comment]
Guys, aren't you gonna marge this already ? This modification is so simple but pretty convenient.

**Comment by Kefta** [comment]
No PRs are being merged for whatever reason:

**Comment by ExtReMLapin** [comment]
this is useless

**Comment by dvdvideo1234** [comment]
What is so useless about it ... Not that I cannot do this myself

**Comment by Acecool** [comment]
Yeah, don't add garbage code ( like the current situation with a lot of the networking functions that aren't type-bound and use string ) which promotes incorrect coding and typing practices ( and the reason the corrections had to be reverted; albeit I wouldn't have reverted the changes because too many old game-modes and addons were incorrectly coded if I was ever in such a situation ).

There are plenty of ways to do many different operations.. if the speed is negligible pick the cleanest one or do testing across more platforms and then choose.

Why not make your case / argument more compelling by providing examples where such a function has been useful to you.

**Comment by dvdvideo1234** [comment]
Abs( ) has many useful features ... but ok....

**Pull request closed without merging**